File: impl/go/large_values_test.go
package jed

// Slice A — out-of-line large values / overflow pages (spec/design/large-values.md §12). A value
// that would push a record past RECORD_MAX spills to a chain of overflow pages (page_type 4),
// leaving a fixed pointer in the record. These per-core tests cover what a static golden cannot
// (an incremental file's bytes depend on commit history): a multi-page chain round-trips, small
// values never spill, the free-list keeps live chains and reclaims dead ones, and the default
// demand-paged file path reads a spilled value back exactly.

import (
	"fmt"
	"path/filepath"
	"strings"
	"testing"
)

// countPageType counts body pages of a given page_type in an image (meta slots start with the magic,
// so they never collide with a small page_type byte).
func countPageType(image []byte, ps int, ty byte) int {
	c := 0
	for i := 0; i+ps <= len(image); i += ps {
		if image[i] == ty {
			c++
		}
	}
	return c
}

// bigValueDB builds an in-memory table with a ~1250-byte text value (forces a multi-page overflow
// chain at page 256: RECORD_MAX = (256-12-12)/2 = 116, cap = 244) plus a small inline value.
func bigValueDB(t *testing.T) (*Database, string) {
	t.Helper()
	db := NewDatabase()
	big := strings.Repeat("abcΩ", 250) // 5 bytes × 250 = 1250 UTF-8 bytes
	mustExec(t, db, "CREATE TABLE t (id int32 PRIMARY KEY, body text)")
	mustExec(t, db, fmt.Sprintf("INSERT INTO t VALUES (1, '%s')", big))
	mustExec(t, db, "INSERT INTO t VALUES (2, 'tiny')")
	return db, big
}

func TestExternalValueSpansOverflowChainAndRoundTrips(t *testing.T) {
	db, big := bigValueDB(t)
	image, err := db.ToImage(256, 1)
	if err != nil {
		t.Fatal(err)
	}
	if n := countPageType(image, 256, pageOverflow); n < 2 {
		t.Fatalf("a large value should span several overflow pages, got %d", n)
	}
	loaded, err := LoadDatabase(image)
	if err != nil {
		t.Fatal(err)
	}
	// Re-serialization is byte-identical (deterministic spill + chain allocation).
	again, err := loaded.ToImage(256, 1)
	if err != nil {
		t.Fatal(err)
	}
	if string(again) != string(image) {
		t.Error("re-serialization of an external value is not byte-identical")
	}
	rows := loaded.RowsInKeyOrder("t")
	if len(rows) != 2 || rows[0][1].Str != big || rows[1][1].Str != "tiny" {
		t.Fatalf("external value did not survive the round trip: %+v", rows)
	}
}

func TestSmallValuesNeverSpill(t *testing.T) {
	db := NewDatabase()
	mustExec(t, db, "CREATE TABLE t (id int32 PRIMARY KEY, v int16)")
	mustExec(t, db, "INSERT INTO t VALUES (1, 10), (2, 20), (3, 30)")
	image, err := db.ToImage(256, 1)
	if err != nil {
		t.Fatal(err)
	}
	if n := countPageType(image, 256, pageOverflow); n != 0 {
		t.Fatalf("inline-fitting values must never spill, got %d overflow pages", n)
	}
}

func TestLoadReclaimsOnlyDeadOverflowPages(t *testing.T) {
	db, _ := bigValueDB(t)
	image, err := db.ToImage(256, 1)
	if err != nil {
		t.Fatal(err)
	}
	loaded, err := LoadDatabase(image)
	if err != nil {
		t.Fatal(err)
	}
	ovf := countPageType(image, 256, pageOverflow)
	if ovf < 2 {
		t.Fatalf("expected a multi-page chain, got %d", ovf)
	}
	// The live value's chain pages are reachable, so they are NOT on the free-list (else a later
	// commit would reuse a still-referenced page).
	if len(loaded.freePages) >= ovf {
		t.Fatalf("live overflow pages (%d) must be reachable, not free (%d free)", ovf, len(loaded.freePages))
	}
}

func TestExternalValueThroughPagedFileAndReclaims(t *testing.T) {
	path := filepath.Join(t.TempDir(), "large_values.jed")
	big := strings.Repeat("Z", 1500) // ≫ RECORD_MAX at ps 256 ⇒ a multi-page overflow chain

	db, err := Create(path, DatabaseOptions{PageSize: 256})
	if err != nil {
		t.Fatal(err)
	}
	mustExec(t, db, "CREATE TABLE t (id int32 PRIMARY KEY, body text)")
	mustExec(t, db, fmt.Sprintf("INSERT INTO t VALUES (1, '%s')", big))
	mustExec(t, db, "INSERT INTO t VALUES (2, 'small')")
	if err := db.Close(); err != nil {
		t.Fatal(err)
	}

	// Reopen demand-paged (the default Open): the big value reconstructs exactly through the
	// pager-backed chain read.
	db, err = Open(path)
	if err != nil {
		t.Fatal(err)
	}
	rows := queryRows(t, db, "SELECT id, body FROM t")
	if len(rows) != 2 || rows[0][1].Str != big || rows[1][1].Str != "small" {
		t.Fatalf("paged read of an external value is wrong: lens %d", len(rows))
	}
	if err := db.Close(); err != nil {
		t.Fatal(err)
	}

	// Delete the big row; its chain is orphaned (leaked this session).
	db, err = Open(path)
	if err != nil {
		t.Fatal(err)
	}
	mustExec(t, db, "DELETE FROM t WHERE id = 1")
	if err := db.Close(); err != nil {
		t.Fatal(err)
	}

	// Reopen: the free-list reconstruction collects only live chains, so the dead chain's pages are
	// now free. Re-inserting a large value reuses them — the high-water grows by a handful of pages,
	// not by a whole fresh chain (~7 pages).
	db, err = Open(path)
	if err != nil {
		t.Fatal(err)
	}
	before := db.pageCount
	mustExec(t, db, fmt.Sprintf("INSERT INTO t VALUES (3, '%s')", big))
	after := db.pageCount
	if after > before+3 {
		t.Fatalf("re-insert did not reuse reclaimed overflow pages (pageCount %d → %d)", before, after)
	}
	if err := db.Close(); err != nil {
		t.Fatal(err)
	}

	// Final correctness through the paged path.
	db, err = Open(path)
	if err != nil {
		t.Fatal(err)
	}
	got := queryRows(t, db, "SELECT body FROM t WHERE id = 3")
	if len(got) != 1 || got[0][0].Str != big {
		t.Fatalf("re-inserted big row is wrong")
	}
	if err := db.Close(); err != nil {
		t.Fatal(err)
	}
}

File: impl/ts/tests/large_values.test.ts
// Slice A — out-of-line large values / overflow pages (spec/design/large-values.md §12). A value
// that would push a record past RECORD_MAX spills to a chain of overflow pages (page_type 4),
// leaving a fixed pointer in the record. These per-core tests cover what a static golden cannot (an
// incremental file's bytes depend on commit history): a multi-page chain round-trips, small values
// never spill, the free-list keeps live chains and reclaims dead ones, and the default demand-paged
// file path reads a spilled value back exactly.

import assert from "node:assert/strict";
import { mkdtempSync, rmSync } from "node:fs";
import { tmpdir } from "node:os";
import { join } from "node:path";
import { test } from "node:test";
import { close, create, Engine, execute, loadEngine, open, toImage } from "../src/tooling.ts";
import { fillerText } from "./util.ts";

const PAGE_OVERFLOW = 4; // page_type for an overflow slab (large-values.md §12)

// countPageType counts body pages of a given page_type in an image (meta slots start with the magic,
// so they never collide with a small page_type byte).
function countPageType(image: Uint8Array, ps: number, ty: number): number {
  let c = 0;
  for (let i = 0; i + ps <= image.length; i += ps) if (image[i] === ty) c++;
  return c;
}

// rowsOf runs a SELECT and returns the rows (asserting it is a query).
function rowsOf(db: Engine, sql: string) {
  const o = execute(db, sql);
  assert.equal(o.kind, "query");
  if (o.kind !== "query") throw new Error("expected a query");
  return o.rows;
}

function textOf(v: { kind: string; text?: string } | undefined): string | null {
  return v && v.kind === "text" ? (v.text ?? null) : null;
}

// A ~1250-byte text value forces a multi-page overflow chain at page 256 (RECORD_MAX = 114, cap = 240).
const BIG = fillerText(1250); // incompressible, so Slice B keeps it external-plain

function bigValueDB(): Engine {
  const db = new Engine();
  execute(db, "CREATE TABLE t (id i32 PRIMARY KEY, body text)");
  execute(db, `INSERT INTO t VALUES (1, '${BIG}')`);
  execute(db, "INSERT INTO t VALUES (2, 'tiny')");
  return db;
}

test("external value spans an overflow chain and round-trips byte-identically", () => {
  const db = bigValueDB();
  const image = toImage(db, 256, 1n);
  assert.ok(
    countPageType(image, 256, PAGE_OVERFLOW) >= 2,
    "a large value spans several overflow pages",
  );

  const loaded = loadEngine(image);
  // Re-serialization is byte-identical (deterministic spill + chain allocation).
  assert.deepEqual(toImage(loaded, 256, 1n), image);
  const rows = rowsOf(loaded, "SELECT id, body FROM t ORDER BY id");
  assert.equal(rows.length, 2);
  assert.equal(textOf(rows[0]![1]), BIG);
  assert.equal(textOf(rows[1]![1]), "tiny");
});

test("small values never spill", () => {
  const db = new Engine();
  execute(db, "CREATE TABLE t (id i32 PRIMARY KEY, v i16)");
  execute(db, "INSERT INTO t VALUES (1, 10), (2, 20), (3, 30)");
  const image = toImage(db, 256, 1n);
  assert.equal(
    countPageType(image, 256, PAGE_OVERFLOW),
    0,
    "inline-fitting values are never externalized",
  );
});

test("load reclaims only dead overflow pages", () => {
  const image = toImage(bigValueDB(), 256, 1n);
  const loaded = loadEngine(image);
  const ovf = countPageType(image, 256, PAGE_OVERFLOW);
  assert.ok(ovf >= 2);
  // Live chain pages are reachable, so they are NOT on the free-list (else a later commit would
  // reuse a still-referenced page).
  assert.ok(
    loaded.freePages.length < ovf,
    `live overflow pages (${ovf}) must not be free (${loaded.freePages.length})`,
  );
});

test("external value through the default demand-paged file path, with reclamation", () => {
  const dir = mkdtempSync(join(tmpdir(), "jed-lv-"));
  const path = join(dir, "large_values.jed");
  const big = fillerText(1500); // incompressible ≫ RECORD_MAX at ps 256 ⇒ a multi-page overflow chain
  try {
    let db = create(path, { pageSize: 256 });
    execute(db, "CREATE TABLE t (id i32 PRIMARY KEY, body text)");
    execute(db, `INSERT INTO t VALUES (1, '${big}')`);
    execute(db, "INSERT INTO t VALUES (2, 'small')");
    close(db);

    // Reopen demand-paged (the default open): the big value reconstructs exactly through the
    // pager-backed chain read.
    db = open(path);
    let rows = rowsOf(db, "SELECT id, body FROM t ORDER BY id");
    assert.equal(rows.length, 2);
    assert.equal(textOf(rows[0]![1]), big);
    assert.equal(textOf(rows[1]![1]), "small");
    close(db);

    // Delete the big row; its chain is orphaned (leaked this session).
    db = open(path);
    execute(db, "DELETE FROM t WHERE id = 1");
    close(db);

    // Reopen: the free-list reconstruction collects only live chains, so the dead chain's pages are
    // now free. Re-inserting a large value reuses them — the high-water grows by a handful of pages,
    // not by a whole fresh chain (~7 pages).
    db = open(path);
    const before = db.pageCount;
    execute(db, `INSERT INTO t VALUES (3, '${big}')`);
    const after = db.pageCount;
    close(db);
    assert.ok(
      after <= before + 3,
      `re-insert did not reuse reclaimed pages (pageCount ${before} → ${after})`,
    );

    // Final correctness through the paged path.
    db = open(path);
    rows = rowsOf(db, "SELECT body FROM t WHERE id = 3");
    assert.equal(rows.length, 1);
    assert.equal(textOf(rows[0]![0]), big);
    close(db);
  } finally {
    rmSync(dir, { recursive: true, force: true });
  }
});
